scratch-debugger/debug.sh

```diff
@@ -0,0 +1,143 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+if [[ $# == 0 ]]; then
+  echo >&2 "USAGE: $0 POD_NAME [POD_NAMESPACE CONTAINER_NAME]"
+  exit 1
+fi
+
+# Customizable parameters
+TMP_SUBDIR="${TMP_SUBDIR:-debug-tools}"
+CONTEXT="${KUBECONTEXT:-}"
+DEBUGGER_NAME="${DEBUGGER_NAME:-debugger}"
+ARCH="${ARCH:-amd64}"
+
+# Name & namespace of target pod
+NAME=$1
+NAMESPACE=${2:-}
+CONTAINER_NAME=${3:-}
+
+# Internal variables
+KUBECTL="kubectl"
+[ -z "$CONTEXT" ] || KUBECTL="$KUBECTL --context=$CONTEXT"
+[ -z "$NAMESPACE" ] || KUBECTL="$KUBECTL --namespace=$NAMESPACE"
+
+TARGET_CNAME=${CONTAINER_NAME}
+if [ -z "$CONTAINER_NAME" ]; then
+  CONTAINER_NAME=$(${KUBECTL} get pod $NAME -o jsonpath='{.spec.containers[0].name}')
+fi
+
+INSTALL_DIR="/tmp/${TMP_SUBDIR}"
+CONTAINER_ID=$(${KUBECTL} get pod $NAME -o jsonpath="{.status.containerStatuses[?(@.name==\"$CONTAINER_NAME\")].containerID}")
+RUNTIME=${CONTAINER_ID%://*}
+CONTAINER_ID=${CONTAINER_ID#*://}
+NODE=$(${KUBECTL} get pod $NAME -o jsonpath='{.spec.nodeName}')
+
+cat <<EOF
+Debug Target Container:
+  Pod:          $NAME
+  Namespace:    ${NAMESPACE:-default}
+  Node:         $NODE
+  Container:    $CONTAINER_NAME
+  Container ID: $CONTAINER_ID
+  Runtime:      $RUNTIME
+
+EOF
+
+if [[ $RUNTIME != docker ]]; then
+  echo >&2 "Error: $0 only works with a docker runtime. Found: $CONTAINER_ID"
+  exit 1
+fi
+
+# Construct the command to debug the target container.
+DEBUG_CMD="${KUBECTL} exec -i -t ${NAME}"
+[ -z "$TARGET_CNAME" ] || DEBUG_CMD="$DEBUG_CMD -c $TARGET_CNAME"
+DEBUG_CMD="$DEBUG_CMD -- ${INSTALL_DIR}/sh -c 'PATH=\$PATH:${INSTALL_DIR} sh'"
+
+if $KUBECTL exec ${NAME} -c $CONTAINER_NAME -- ${INSTALL_DIR}/echo &>/dev/null; then
+  echo "Debug tools already installed. Dumping you into the pod container now."
+  eval "$DEBUG_CMD"
+  exit 0
+fi
+
+echo "Installing busybox to $INSTALL_DIR ..."
+
+# Cleanup the debugger pod.
+function cleanup() {
+  if ${KUBECTL} get pod ${DEBUGGER_NAME} &>/dev/null; then
+    ${KUBECTL} delete pod ${DEBUGGER_NAME}
+  fi
+}
+
+case $ARCH in
+  "amd64")
+    IMAGE=busybox
+    ;;
+  "arm")
+    IMAGE=armhf/busybox
+    ;;
+  "arm64")
+    IMAGE=aarch64/busybox
+    ;;
+  "ppc64le")
+    IMAGE=ppc64le/busybox
+    ;;
+  "s390x")
+    IMAGE=s390x/busybox
+    ;;
+  *)
+    echo "Unsupported architecture: $ARCH"
+    exit 1
+esac
+
+# If the debugger is not already running...
+if ! ${KUBECTL} get pod ${DEBUGGER_NAME} &>/dev/null; then
+  # Start the debug pod
+  cat ${BASH_SOURCE[0]%/*}/debugger.yaml | \
```

Comment:
Suggestion: it might be nice to inline debugger.yaml so the script is stand-alone. I suspect this might see use outside of Kubernetes developers.